blockly/blocks/component_variable.js

```diff
@@ -0,0 +1,172 @@
+ * @param {!Blockly.Block} block The block containing this text.
+ */
+Blockly.Blocks.ComponentFieldVariable.prototype.init = function(block) {
+  if (this.sourceBlock_) {
+    // Dropdown has already been initialized once.
+    return;
+  }
+  Blockly.Blocks.ComponentFieldVariable.superClass_.init.call(this, block);
+};
+
+Blockly.Blocks.ComponentFieldVariable.ComponentVariables = function(root, component_type) {
+  var blocks;
+  if (root.getDescendants) {
+    // Root is Block.
+    blocks = root.getDescendants();
+  } else if (root.getAllBlocks) {
+    // Root is Workspace.
+    blocks = root.getAllBlocks();
+  } else {
+    throw 'Not Block or Workspace: ' + root;
+  }
+  var variableHash = Object.create(null);
+  // Iterate through every block and add each variable to the hash.
+  for (var x = 0; x < blocks.length; x++) {
+    if (blocks[x].getComponentName && blocks[x].getVars && 
+        (blocks[x].getComponentName() ==  component_type)) {
+      var blockVariables = blocks[x].getVars();
+      for (var y = 0; y < blockVariables.length; y++) {
+        var varName = blockVariables[y];
+        // Variable name may be null if the block is only half-built.
+        if (varName) {
+          variableHash[varName.toLowerCase()] = varName;
+        }
+      }
+    }
+  }
+  // Flatten the hash into a list.
+  var variableList = [];
+  for (var name in variableHash) {
+    variableList.push(variableHash[name]);
+  }
+  return variableList;
+};
+
+
+/**
+ * Return a sorted list of variable names for variable dropdown menus.
+ * Include a special option at the end for creating a new variable name.
+ * @return {!Array.<string>} Array of variable names.
+ * @this {!Blockly.FieldVariable}
+ */
+Blockly.Blocks.ComponentFieldVariable.dropdownCreateComponents = function() {
+  if (this.sourceBlock_ && this.sourceBlock_.workspace) {
+    var variableList =
+        Blockly.Blocks.ComponentFieldVariable.ComponentVariables(this.sourceBlock_.workspace, 
+                this.component_type);
+  } else {
+    var variableList = [];
+  }
+  // Ensure that the currently selected variable is an option.
+  var name = this.getText();
+  if (name && variableList.indexOf(name) == -1) {
+    variableList.push(name);
+  }
+  variableList.sort(goog.string.caseInsensitiveCompare);
+  variableList.push(Blockly.Msg.RENAME_VARIABLE);
+  variableList.push(Blockly.Msg.NEW_VARIABLE);
+  // Variables are not language-specific, use the name as both the user-facing
+  // text and the internal representation.
+  var options = [];
+  for (var x = 0; x < variableList.length; x++) {
+    options[x] = [variableList[x], variableList[x]];
+  }
+  return options;
+};
+
+
+/**
+ * Finds all user-created instances of the ComponentFieldVariable block config.
+ * @return {!Array.<string>} Array of instance names.
+ */
+Blockly.Blocks.ComponentFieldVariable.Instances = function(component_type) {
+  var instList = [];
+  var blocks = Blockly.mainWorkspace.getAllBlocks();
```

Comment:
`Blockly.Blocks.ComponentFieldVariable.CheckSetupPresent` should be passed an instance of the workspace to pass to this function.